### ISSUE TITLE
Avoid loading new assemblies into the VS process

### ISSUE THREAD

**Issue by mikkelbu** [nunit/nunit.analyzers#196]
Issue raised based on comments from @jnm2 and @JohanLarsson in https://gitter.im/GuOrg/Gu.Roslyn.Asserts?at=5e793252b08f3e0abf8c2224.

So instead of loading types and possibly new assemblies into the VS process we handle intrinsic type converters directly and will handle custom type converters in a subsequent issue - #197. If a converter (from string) exists then we assume that it is able to convert to the target type. We loose the ability to mark actual string values that does not satisfy the correct format as failures, but that is the price to pay using static analysis.

Extract from Gitter.
> This is not a good idea:
https://github.com/nunit/nunit.analyzers/blob/master/src/nunit.analyzers/Extensions/AttributeArgumentSyntaxExtensions.cs#L154
this could load new assemblies into the VS process
Anything in the GAC or VS directories
...

**Comment by Dreamescaper** [comment]
I think we can leave Convert.ChangeType for some limited set of types (e.g. for Uri) - without trying to resolve types dynamically or handle custom converters.

**Comment by Dreamescaper** [comment]
E.g. something like that:
```csharp
        private static IReadOnlyList<Type> ConvertibleTypes = new List<Type> { 
            typeof(Uri), 
            typeof(DateTimeOffset),
			...
        };

        private static bool TryChangeType(ITypeSymbol targetTypeSymbol, object argumentValue)
        {
            var typeName = targetTypeSymbol.GetFullMetadataName();
            var targetType = ConvertibleTypes.FirstOrDefault(t => t.FullName == typeName);
            
            if (targetType == null)
            {
                return false;
            }

            try
            {
                Convert.ChangeType(argumentValue, targetType, CultureInfo.InvariantCulture);
                return true;
            }
            catch { 
                ...
            }
        }
```

**Comment by mikkelbu** [comment]
Thanks for the suggestion @Dreamescaper. I'll make another PR and see how it looks (and behaves), but it will probably be tomorrow, as I'm heading to bed now 😄.

**Comment by jnm2** [comment]
`Convert.ChangeType` is fine because it doesn't load user code into the analysis process. It wouldn't be a great idea to use it if it behaves differently between .NET Framework and .NET Core or between two versions of .NET Core, because then you get different diagnostics depending on whether the analyzer is running in VS or from the command line (etc). Just remember that the way that `Convert.ChangeType` behaves for the analyzer doesn't really represent how it behaves when the user's test project runs, unless `Convert.ChangeType` behaves exactly the same in all frameworks forever.

**Comment by mikkelbu** [comment]
It does not seem like `Convert.ChangeType` is invoking the converters. If I run
```csharp
        [Test]
        public void TestTimeSpan()
        {
            var input = "23:59:59";
            var result = Convert.ChangeType(input, typeof(TimeSpan), CultureInfo.InvariantCulture);
        }

        [Test]
        public void TestDateTimeOffset()
        {
            var input = "2019-10-14T19:15:25+00:00";
            var result = Convert.ChangeType(input, typeof(DateTimeOffset), CultureInfo.InvariantCulture);
        }
```
then both tests fail with exceptions like
```
  Message: 
    System.InvalidCastException : Invalid cast from 'System.String' to 'System.DateTimeOffset'.
  Stack Trace: 
    Convert.DefaultToType(IConvertible value, Type targetType, IFormatProvider provider)
    IConvertible.ToType(Type type, IFormatProvider provider)
    Convert.ChangeType(Object value, Type conversionType, IFormatProvider provider)
```
as none of `IConvertible` methods can be used to convert to a `DateTimeOffset` (or `TimeSpan`, respectively). 

Since I know that the source type will be either `double`, `string`, or `int` and that the target type will also be in a fixed set of types (`DateTimeOffset`, `TimeSpan`, `Guid`, `Uri`, and `Version`), can I then still use the following case for these types, or is is possible to override the default converters
```csharp
                    TypeConverter converter = TypeDescriptor.GetConverter(reflectionTargetType);
                    if (converter.CanConvertFrom(reflectionArgumentType))
                    {
                        try
                        {
                            converter.ConvertFrom(null, CultureInfo.InvariantCulture, argumentValue);
                            return true;
                        }
                        catch
                        {
                            return false;
                        }
                    }
```
Ps. If the above does not make sense I can provide a complete PR, if that is easier to talk about

**Comment by Dreamescaper** [comment]
Since the idea is to have limited amount of supported types, I would suggest to actually _use_ default converters (e.g. `UriTypeConverter`, `VersionConverted`, `GuidConverter`, `DateTimeOffsetConverter`), and drop `TypeDescriptor.GetConverter` call altogether. 
Seems like a safer way for me.

**Comment by mikkelbu** [comment]
I think you are right. Thanks for the suggestion 👍.

Ps. I guess it is not possible to override these providers, as it would require us to decorate system types with new attributes.

**Comment by Dreamescaper** [comment]
There is method `TypeDescriptor.AddProvider` which may do something bad in our case, not sure.

**Comment by jnm2** [comment]
`Convert.ChangeType` does not use type converters. I read that it was originally added as a stopgap to make .NET easier to use for VB developers and was considered legacy in .NET Framework 1.0. Type converters were going to be the way forward.

**Comment by jnm2** [comment]
> `VersionConverter`

Btw: VersionConverter does not exist in .NET Framework or in .NET Core prior to 3.0.
https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.versionconverter?view=netcore-3.1#moniker-applies-to
